### ISSUE TITLE
Feature proposal: Batch windows for semi-incremental learning

### ISSUE THREAD

**Issue by Dandandan** [creme-ml/creme#134]
Hi, I saw your presentation online @MaxHalford and am thinking of contributing to this library.

What I am proposing is to add a feature to creme that keeps a window in memory of a fixed number of data samples.

*Advantages*
* Possibly higher accuracy (can fit multiple iterations using same data within window)
* Possible to use mini-batches for vectorization, data parallelism, GPU, etc. implementations
* Still using constant amount of memory, feeding individual samples
* Could be used for a large range of models
* Possible to implement features requiring mini-batches, such as batch normalization

*Disadvantages*
* Higher memory use
* More computation per `fit_one`
* Additional parameters to set / finetune

Open questions:
* Does this still fit into the idea of incremental learning, or is it something you'd like to avoid
* Any references available / papers / other implementations that use this idea?
* When to start fitting the underlying model, already after the first sample or after we reach `mini_batch_size` size?
* Should we have one bigger buffer (e.g. of 1024 samples) and one mini-batch size to train on? In this setting we could e.g. shuffle the samples in the window for (possibly) better generalization like with batch learning
* Any other ideas / feedback?

Implementation
===
What I am thinking about is to create a model wrapper that will work on multiple models.

pseudo implementation:

```python
class BatchWindowModel:
    def __init__(self, model, batch_window_size, mini_batch_size=32):
        self.model = model
        self.batch_window_size = batch_window_size
        self.batch_window = collections.deque()
        self.mini_batch_size = mini_batch_size

    def fit_one(self, x, y):
        if len(self.batch_window) > self.batch_window_size:
            self.batch_window.popleft()

        self.batch_window.append((x, y))

        for (a, b) in self.batch_window:
            self.model.fit_one(a, b)

    def predict_one(self, x):
        return self.model.predict_one(x)
```

usage example:

```
lin_reg = linear_model.LinearRegression()
lin_reg2 = linear_model.LinearRegression()

batch_model = BatchWindowModel(lin_reg2, 32)
metric1 = metrics.MAE()
metric2 = metrics.MAE()

for x_rest, y in datasets.fetch_restaurants():
    x = dict()
    x['weekday'] = x_rest['date'].weekday()
    x['is_weekend'] = x_rest['date'].weekday() in (5, 6)
    y_pred = lin_reg.predict_one(x)

    lin_reg.fit_one(x, y)

    metric1.update(y, y_pred)

for x_rest, y in datasets.fetch_restaurants():
    x = dict()
    x['weekday'] = x_rest['date'].weekday()
    x['is_weekend'] = x_rest['date'].weekday() in (5, 6)
    y_pred = batch_model.predict_one(x)

    batch_model.fit_one(x, y)

    metric2.update(y, y_pred)

print(metric1)
print(metric2)

```

**Comment by MaxHalford** [comment]
Hey!

Thanks for your interest in the library! Things are a bit slow at the moment because most of the people working on `creme` are on holiday this summer. 

Your proposal is in the right direction. The thing is that not much research has been done on this topic, at least from what I've seen. However I'm quite sure that there is something to gain from storing previous examples and retraining on them.

Your proposal is interesting, but if I may say so I believe you could do something a bit smarter. Indeed you could store a window of `k` examples with the highest errors. Then, every so often you could take one of these worst examples and call `fit_one` on it. In other words you would be doing some kind of oversampling. I believe that this should be better because effectively you're asking the model to retrain on examples where it performed badly, and you're discarding "easy" examples,  and thus saving some computing power. Also you don't have to change much of your code to make this work (I think).

I don't have time to code a proof of concept because I'm on holiday, sorry! But feel free to ask as many questions as you wish :)

PS: it would be nice to some research on "online oversampling/undersampling". Also if we go down this way we could call the associated module `imblearn`.

**Comment by Dandandan** [comment]
Thanks for the feedback, interesting idea. Indeed, that makes a lot of sense and might work for a lot of datasets. 
I also did something similar some time ago and that worked pretty well (only training on x% with highest error). It probably might not work as well for datasets with a high amount of label noise / wrongly assigned labels.

I will come up with an implementation for your proposal and a few experiments when I get to this.

**Comment by Dandandan** [comment]
Here a PoC for the error-based window (without discarding everything besides k worst examples):

```
class BatchWindowModelError:
    def __init__(self, model, batch_window_size, error_metric, batch_size):
        self.model = model
        self.batch_window = queue.PriorityQueue()
        self.error_metric = error_metric
        self.batch_size = batch_size

    def fit_one(self, x, y):
        examples = []
        self.batch_window.put((-math.inf, (x, y)))
        for _ in range(self.batch_size):
            if self.batch_window.empty():
                break
            (_, (a, b)) = self.batch_window.get()
            self.model.fit_one(a, b)
            y_pred = self.predict_one(a)

            error = self.error_metric(b, y_pred)
            examples.append((-error, (a, b)))

        for e in examples:
            self.batch_window.put(e)

    def predict_one(self, x):
        return self.model.predict_one(x)
```

For the example above it doesn't work very well (MAE 81.282155 vs 12.808271).

**Comment by MaxHalford** [comment]
I just stumbled on [this](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.144.2279&rep=rep1&type=pdf) paper which seems slightly related.

**Comment by johny-c** [comment]
Hi all, 
@MaxHalford thanks for your effort, this is a useful library. 
@Dandandan, your idea would indeed be very useful. 
I wanted to mention that the ideas proposed here are similar to ideas used in modern Reinforcement Learning. There learning **has** to happen online as an agent interacts with a (usually simulated) environment. The idea of a fixed memory size is equivalent to an [Experience Replay Buffer](https://apps.dtic.mil/dtic/tr/fulltext/u2/a261434.pdf) and @MaxHalford 's idea of retraining on high-error samples is pretty much equivalent to [Prioritized Experience Replay](https://arxiv.org/pdf/1511.05952.pdf).
I just wanted to mention these things in case you are not aware, so you don't do some double work and also so you can perhaps get some ideas to implement that could also benefit the RL community.
Cheers ;)

**Comment by MaxHalford** [comment]
That's some really nice information @johny-c. Putting names on concepts is very useful, especially to me.

**Comment by Dandandan** [comment]
Thanks all for the pointers... I also think some research into existing strategies & algorithms is needed and huge thanks @johny-c for the pointers into methods used in RL... Maybe we also can re-use the correct terminology...

**Comment by MaxHalford** [comment]
Hey @Dandandan, just wanted to let you know I just added an `imblearn` module for imbalanced learning. Maybe that we could add a class which implements your idea. I like the idea of calling it `ExperienceReplay` or `Replayer`.

**Comment by Dandandan** [comment]
Cool, yeah sure would love to help.
Probably have some time next week, early January.

**Comment by MaxHalford** [comment]
No worries. I'm in the process of figuring out formulas for random oversampling. I'll be adding it to `creme` by the end of the week. I'll be updating [this](https://github.com/creme-ml/creme/blob/master/docs/notebooks/imbalanced-learning.ipynb) notebook for a comparison of different over/under-sampling methods.

**Comment by Dandandan** [comment]
Interesting! I may have some time to look into this tomorrow...

Generally I think we need to have a very basic implementation, e.g. a fixed replay buffer, with a simple sampling strategy (e.g. uniform sampling), that could be extended easily.

It would be interesting to see what it does for imbalanced datasets as well...

**Comment by MaxHalford** [comment]
Okay I have a bit of news: I've tried to implement experience replay by allowing the samplers in `imblearn` to run from the observations in a random order. Here is an example using `RandomOverSampler`:

```python
class Shuffler(collections.UserList):

    def __init__(self, buffer_size, seed=None):
        super().__init__()
        self.buffer_size = buffer_size
        self.seed = seed
        self.rng = random.Random(seed)

    def push(self, el):

        if self.buffer_size == 0:
            return el

        if len(self) == self.buffer_size:
            i = self.rng.randint(0, len(self) - 1)
            out = self[i]
            self[i] = el
            return out

        self.append(el)
        return None

class RandomOverSampler(base.Sampler):

    def __init__(self, classifier, desired_dist, seed=None):
        super().__init__(classifier=classifier, seed=seed)
        self.desired_dist = desired_dist
        self._actual_dist = collections.Counter()
        self._pivot = None
        self.shuffler = Shuffler(buffer_size=32, seed=seed)

    def fit_one(self, x, y):

        self._actual_dist[y] += 1
        f = self.desired_dist
        g = self._actual_dist

        # Check if the pivot needs to be changed
        if y != self._pivot:
            self._pivot = max(g.keys(), key=lambda y: g[y] / f[y])
        else:
            xy = self.shuffler.push((x, y))
            if xy is not None:
                self.classifier.fit_one(*xy)
            return self

        M = g[self._pivot] / f[self._pivot]
        rate = M * f[y] / g[y]

        for _ in range(self._rng.poisson(rate)):
            xy = self.shuffler.push((x, y))
            if xy is not None:
                self.classifier.fit_one(*xy)

        return self
```

The thing is that it doesn't really help... At least on the examples I tried. Basically oversampling means that the model learns successively on the same sample. I thought that randomizing the training order would help, but it doesn't seem so. I probably have to do some more experiments but overall it doesn't seem like experience replaying is worthwhile.

Prioritized experience replay, on the other hand, seems like a really good idea to try out, and I think we should head in that direction. As for your original proposal @Dandandan, I hope you can agree that it's interesting but probably needs more significant score gains before we can put into `creme`. I rather only keep the algorithms that bring something significant to the table.

**Comment by MaxHalford** [comment]
@raphaelsty has implemented `sampling.HardSamplingRegressor` and `sampling.HardSamplingClassifier` to do prioritized experience replay. I think we can now close this issue :). There's always room for improvement but we've made very progress.

**Comment by Dandandan** [comment]
Sounds good @MaxHalford ! Did also try out a similar approach around December, had about the same experience on some datasets... Looking forward to try the prioritized experience replay functionality!

**Comment by MaxHalford** [comment]
Please do! @raphaelsty is busy at the moment, but I know for sure that he's interested in pursuing this direction as it's giving good results.
